server/comms/server.go

```diff
@@ -380,6 +478,108 @@ func (s *Server) banish(ip dex.IPKey) {
 	s.quarantine[ip] = time.Now().Add(banishTime)
 }
 
+// wsLimiter gets any existing routeLimiter for an IP incrementing the
+// connection count for the address, or creates a new one. The caller should use
+// wsLimiterDone after the connection that uses routeLimiter is closed. Loopback
+// addresses always get a new unshared limiter. NewIPKey should be used to
+// create an IPKey with interface bits masked out. This is not perfect with
+// respect to remote IPv6 hosts assigned multiple subnets (up to 16 bits worth).
+// Disable IPv6 if this is not acceptable.
+func (s *Server) wsLimiter(ip dex.IPKey) *routeLimiter {
+	// If the ip is a loopback address, this likely indicates a hidden service
+	// or misconfigured reverse proxy, and it is undesirable for many such
+	// connections to share a common limiter. To avoid this, return a new
+	// untracked limiter for such clients.
+	if ip.IsLoopback() {
+		return newRouteLimiter()
+	}
+
+	s.wsLimiterMtx.Lock()
+	defer s.wsLimiterMtx.Unlock()
+	prefix := ip.PrefixV6()
+	if prefix != nil { // not ipv4
+		if n := s.v6Prefixes[*prefix]; n > 0 {
+			log.Infof("Detected %d active IPv6 connections with same prefix %v", n, prefix)
+			// ip = *prefix // Use a prefix-aggregated limiter.
+		}
+	}
+
+	if l := s.wsLimiters[ip]; l != nil {
+		if l.conns >= rpcMaxConnsPerIP {
+			return nil
+		}
+		l.conns++
+		if prefix != nil {
+			s.v6Prefixes[*prefix]++
+		}
+		if l.cleaner != nil { // l.conns was zero
+			log.Debugf("Restoring active rate limiter for %v", ip)
+			// Even if the timer already fired, we won the race to the lock and
+			// incremented conns so the cleaner func will be a no-op.
+			l.cleaner.Stop() // false means timer fired already
+			l.cleaner = nil
+		}
+		return l.routeLimiter
+	}
+
+	limiter := newRouteLimiter()
+	s.wsLimiters[ip] = &ipWsLimiter{
+		conns:        1,
+		routeLimiter: limiter,
+	}
+	if prefix != nil {
+		s.v6Prefixes[*prefix]++
+	}
+	return limiter
+}
+
+// wsLimiterDone decrements the connection count for the IP address'
+// routeLimiter, and deletes it entirely if there are no remaining connections
+// from this address.
+func (s *Server) wsLimiterDone(ip dex.IPKey) {
+	s.wsLimiterMtx.Lock()
+	defer s.wsLimiterMtx.Unlock()
+
+	if prefix := ip.PrefixV6(); prefix != nil {
+		switch s.v6Prefixes[*prefix] {
+		case 0:
+		case 1:
+			delete(s.v6Prefixes, *prefix)
+		default:
+			s.v6Prefixes[*prefix]--
+		}
+	}
+
+	wsLimiter := s.wsLimiters[ip]
+	if wsLimiter == nil {
+		return // untracked limiter
+		/*
+			// Try a prefix-aggregated limiter.
+			if prefix == nil {
+				return
+			}
+			wsLimiter = s.wsLimiters[*prefix]
+			if wsLimiter == nil {
+				return
+			}
+			ip = *prefix
+		*/
```

Comment:
Intended to leave in?

Would an untracked limiter be expected? Are these the loopback addresses?

Reply:
This code would pair with the commented line in `wsLimiter` that switches the connection to a "prefix-aggregated" limiter.  https://github.com/decred/dcrdex/pull/1192/files#diff-5e901c2a3618ece1ff4d4750b55e2d616dded4726f9258f4043c994c6f41b390R503

I have this all commented because it's an incomplete approach, since when it does detect the need to aggregate, it leaves existing connections from the same prefix with their individual limiters.  It would probably need to replace their limiters with the aggregated one to be most effective.  The topic of IPv6 rate limiting is very much unsolved, but the best suggestion I've seen so far is probably https://serverfault.com/a/919324/190378

It's just experimental, so I'll remove this block.  Presently the only purpose of that `v6Prefixes` map is the `log.Debugf` line so we can learn in the field.

Reply:
> Would an untracked limiter be expected? Are these the loopback addresses?

Right.